telosnetwork/teloscan
# #380 | CSV Export

TODO discuss adding CSV export endpoint / hcaptcha validation endpoint

# Fixes #380 

## Description
This PR adds the ability to export transactions and token transfers as CSV

## Test scenarios
- run this branch locally: run `git fetch --all && git checkout 380-csv-export && yarn install && yarn dev`
- go to http://localhost:8080
- open the 'Advanced' menu on the home page
    - you should see a new Export Data option
- click Export Data
    - you should be taken to the new export page
- go to any account page, e.g. http://localhost:8080/address/0x160505F3dfD1cb58B91e322c828Ae0F74c043c3C
- scroll to the bottom
    - you should see a 'Download CSV' link
- click the download link
    - you should be taken to the new export page with the address pre-filled and the 'transactions' option selected
- go to an erc-20 transfers page, e.g. http://localhost:8080/address/0x160505F3dfD1cb58B91e322c828Ae0F74c043c3C#erc20_transfers
- scroll to the bottom
    - you should see a 'Download CSV' link
- click the download link
    - you should be taken to the new export page with the address pre-filled and the 'ERC-20 Transfers' option selected
- go to an erc-721 transfers page, e.g. http://localhost:8080/address/0x160505F3dfD1cb58B91e322c828Ae0F74c043c3C#erc721_transfers
- scroll to the bottom
    - you should see a 'Download CSV' link
- click the download link
    - you should be taken to the new export page with the address pre-filled and the 'ERC-721 Transfers' option selected
- go to an erc-1155 transfers page, e.g. http://localhost:8080/address/0x160505F3dfD1cb58B91e322c828Ae0F74c043c3C#erc1155_transfers
- scroll to the bottom
    - you should see a 'Download CSV' link
- click the download link
    - you should be taken to the new export page with the address pre-filled and the 'ERC-1155 Transfers' option selected
- set Export Type to `Transactions`
- enter a date range
    - the Download button should still be disabled
- complete the captcha
    - the Download button should be enabled
- click the Reset button
    - form fields should clear
- re-enter an address and date range
- click Download
    - the download button should enter a loading state
    - the data should download
    - the file name should make sense
    - the rows should be correct (make sure to check the details of a couple transactions against teloscan)
    - after the data downloads, the button should no longer be in a loading state and a notification should pop
- open the downloaded file in a program which can view CSV data like google sheets, or in VSCode using the extension Rainbow CSV
    - the file should be correctly formatted
- go back to the download page
- switch export type to ERC-20 Transfers
- click Download
    - the download button should enter a loading state
    - the data should download
    - the file name should make sense
    - the rows should be correct (make sure to check the details of a couple transactions against teloscan)
    - after the data downloads, the button should no longer be in a loading state and a notification should pop
- verify the downloaded file is formatted correctly
- repeat the last 3 steps for ERC-721 and ERC-1155 transfers
- change the download option to Block Range
- enter a block range (if using the account above, `0x...3c3c`, select Transactions and then you can enter start block `320792430` to  end block `320792434` to get one tx)
    - you should see the proper number of transactions for the block range
- using the selector in the Advanced menu, change the language to each language listed
    - the page should be fully localized





## Checklist:
